Inline several functions from util.c

khwilliamson opened the pull request:
These trivial functions will likely get optimized out.  But I'm unsure if inlining all of them is a good idea

khwilliamson committed:
Inline get_vtbl()

This trivial function will likely get optimized out

tonycoz commented on inline.h:
```diff
@@ -3698,6 +3740,46 @@ Perl_my_strlcpy(char *dst, const char *src, Size_t size)
+/*
+
+=for apidoc sv_nosharing
+
+Dummy routine which "shares" an SV when there is no sharing module present.
+Or "locks" it.  Or "unlocks" it.  In other
+words, ignores its single SV argument.
+Exists to avoid test for a C<NULL> function pointer and because it could
+potentially warn under some level of strict-ness.
+
+=cut
+*/
+
+PERL_STATIC_INLINE void
+Perl_sv_nosharing(pTHX_ SV *sv)
+{
+    PERL_UNUSED_CONTEXT;
+    PERL_UNUSED_ARG(sv);
+}
+
+/*
+
+=for apidoc sv_destroyable
+
+Dummy routine which reports that object can be destroyed when there is no
+sharing module present.  It ignores its single SV argument, and returns
+'true'.  Exists to avoid test for a C<NULL> function pointer and because it
+could potentially warn under some level of strict-ness.
+
+=cut
+*/
+
+PERL_STATIC_INLINE bool
+Perl_sv_destroyable(pTHX_ SV *sv)
+{
+    PERL_UNUSED_CONTEXT;
+    PERL_UNUSED_ARG(sv);
+    return TRUE;
+}
+
```
These are only called via function pointers, there's no point to inlining them

khwilliamson replied:
PR updated to not include these two.

tonycoz commented on embed.fnc:
```diff
@@ -627,7 +627,7 @@ CRTopd	|Malloc_t|realloc	|Malloc_t where 			\
 CTopd	|Free_t |mfree		|Malloc_t where
 
-CTpdR	|void * |get_context
+CTsdR	|void * |get_context
```
Shouldn't these have the `i` flag?

khwilliamson replied:
Yes

tonycoz commented on inline.h:
```diff
@@ -3624,6 +3624,48 @@ Perl_savesharedsvpv(pTHX_ SV *sv)
     return savesharedpvn(pv, len);
 }
 
+#ifndef PERL_GET_CONTEXT_DEFINED
+
+/*
+=for apidoc_section $embedding
+=for apidoc get_context
+
+Implements L<perlapi/C<PERL_GET_CONTEXT>>, which you should use instead.
```
> ..\av.c(1208): error C2129: static function 'void *Perl_get_context(void)' declared but not defined

The version from `win32/win32thread.c` should probably be inlined too, possibly in `win32/win32thread.h`

jkeenan commented:
> These trivial functions will likely get optimized out. But I'm unsure if inlining all of them is a good idea

@khwilliamson, this pull request is failing to compile during `make` on Win32.  Could you investigate?

khwilliamson committed:
embed.fnc: Mark some inline functions as such

I had thought that the 's' flag was sufficient for making a function
inline.  But because of PERL_NO_INLINE_FUNCTIONS, the 'i' flag is also
needed.

demerphq commented:
'f' has to be the most minimal commit message I have seen so far. ;-)

Typo?

demerphq commented on embed.fnc:
```diff
@@ -529,7 +529,9 @@
 :	 This is used for functions that have always had a Perl_ prefix, but
-:	 have been moved to a header file and declared static.
+:	 have been moved to a header file and declared static.  If those
+:	 functions are also inlined, the 'i' or 'I' flags should also be
```
FWIW, this is not correct. the 's', 'S', 'i' and 'I' (capital I) flags are mutually exclusive. You can only have one. I think you have been mislead as there is a bug in embed.pl that uses the first one in the flags definition. I just happened to stumble on this bug myself independently of this ticket, and will push a PR to fix it in the next minutes. Note the code that looks like this: 

```
                $type = {
                    'S' => 'STATIC',
                    's' => 'STATIC',
                    'i' => 'PERL_STATIC_INLINE',
                    'I' => 'PERL_STATIC_FORCE_INLINE'
                }->{$1};
```

demerphq replied:
See https://github.com/Perl/perl5/pull/20755

khwilliamson committed:
Inline get_context() for non-Win32

This trivial function should get optimized out.  But I couldn't get it
to work for Windows, because the two likely hdr files don't have
PL_thr_key defined in them.  I suppose a new hdr file could be created
that gets included later.  But I didn't think it was worth it.

tonycoz commented on embed.fnc:
```diff
@@ -6108,11 +6107,13 @@ Cp	|int	|do_spawn	|NN char *cmd
 Cp	|int	|do_spawn_nowait|NN char *cmd
 #endif /* defined(VMS) || defined(WIN32) */
 #if defined(WIN32)
+CTsdR	|void * |get_context
```
I suspect the Win32 failures are because the Win32 Perl_get_context() isn't static.

demerphq replied:
FWIW, I fixed that as part of the conflict resolution on embed.fnc

demerphq commented:
@khwilliamson I took the liberty of resolving the embef.fnc conflicts and resolved the win32 get_context not being static that was  raised by @tonycoz above at the same time and forced pushed.

demerphq reviewed (commented):
Looks ok to me now, but lets see how the tests work out.

demerphq reviewed (approved):
Approving to clear the "request changes" status.
